### ISSUE TITLE
A minimal custom transformer plugin proposal

### ISSUE THREAD

**Issue by jakebailey** [microsoft/TypeScript#54276]
In the TypeScript compiler, we have the concept of a "transformer". A transformer is a function which "transforms" one AST to another. TypeScript ships many such transforms, including all of the various transformation steps needed to downlevel newer syntax to older syntax, to convert from `import`/`export` to `require`/`module.exports`, and so on. Even the declaration emitter is a transformer, stripping away function bodies, filing in types, etc.

Since TypeScript 2.3, TypeScript has had the following API:

```ts
interface Program {
    emit(
        targetSourceFile?: SourceFile,
        writeFile?: WriteFileCallback,
        cancellationToken?: CancellationToken,
        emitOnlyDtsFiles?: boolean,
        customTransformers?: CustomTransformers, // <--- 👀
    ): EmitResult;
}

interface CustomTransformers {
    before?: (TransformerFactory<SourceFile> | CustomTransformerFactory)[];
    after?: (TransformerFactory<SourceFile> | CustomTransformerFactory)[];
    afterDeclarations?: (TransformerFactory<Bundle | SourceFile> | CustomTransformerFactory)[];
}
```

This API allows users to provide a set of "custom transformers" at `emit` time. These transformers run alongside TypeScript's own transformers, either before or after, depending on where they are placed in `CustomTransformers`.

However, while this API exists and does work, a major gotcha is that custom transformers can only be provided to the _API_. There is no way to use custom transformers if you want to use `tsc` alone.

This has been a long-standing issue; [#14419](https://github.com/microsoft/TypeScript/issues/14419) is the fourth most upvoted issue on the repo.

# What do people use custom transformers for?

This list can go on and on, but transformers are commonly used these days for:

- Generating serialization / deserialization / runtime type checks
  - `typia`,`ts-runtime-checks`, which top the [benchmark charts](https://moltar.github.io/typescript-runtime-type-benchmarks/), but there are many more
- Powering internationalization
  - `formatjs`
- Generating database code
  - `ts-graphql-plugin`
- Transforming paths (please don't do this one, though 😅)

I also know of many ideas which _want_ to be implemented via plugins, but are not willing to do so without official support, including other runtime type checks, tracing code, and so on.

# How do users currently use custom transformers?

Given the current constraints of custom transformers, there are two ways that downstream users have been able to use them:

1. Build TypeScript code via the API, either directly or by using some system which uses the API.
   - `webpack`, `rollup`'s TS plugins, `nx`, all provide configuration which allows users to specify a set of custom transformers.
1. Patch TypeScript.
   - `ttypescript`, `ts-patch` wholly replace TypeScript in a user's workflow.

The first one is obviously fine. The latter, eek!

# What changed?

In TypeScript 5.0, we shipped the biggest architectural / packaging change in TypeScript's history; conversion of the entire project to modules, including the bundling of our code. The relevant side-effects are twofold:

1. The structure of `tsc.js` (and other outputs) changed massively. This broke anyone who was patching the contents of the TypeScript package.
1. The API is no longer runtime patchable; we use `esbuild` to bundle our code and the objects it creates are not "configurable", correctly emulating how "real" ES modules would behave. Any runtime patches were already extremely fragile, only functioning due to the structure of our old `namespace` emit.

In response to this, I did [an analysis](https://gist.github.com/jakebailey/3e73c7aab4da2044a95121467aa3b4e9) of as many TS "patchers" as I could find, and found that there are actually very few reasons to patch TypeScript at all:

- `ttypescript`/`ts-patch`, which seem to be almost exclusively used to enable the use of custom transformers.
- Language service plugins, mainly injecting their own module resolution and custom source file types (`vue`/`volar`, `css`, etc).
- Build systems like `heft`.
- Yarn PnP.
- Those who want to tree shake out our parser (`prettier`).

For 5.0, I was able to fix many of these projects (or at least guide / warn maintainers ahead of time), but many of these patchers have no viable alternative.

We have been seriously considering ways that we can approach the problems that remain in hopes that we can eliminate the need for people to patch TypeScript. This includes (no promises):

- Custom transformers.
- Custom module resolution.
- A future API that actually allows a consumer to tree shake out _just_ a parser (or similar).

This issue intends to address the first bullet.

# A conservative, targeted proposal

In short, my proposal is to add to TypeScript the ability to define "custom transformer plugins". This new functionality is _targeted and minimal_. Its only intent is to satisfy those who just want to be able to add custom transformers.

Many details still have yet to be expanded on, but in short:

- These plugins are placed into the `plugins` array of `compilerOptions`, with a discriminator like `"type": "transformer"`.
  - This is the approach taken by `ttypescript`/`ts-patch`, and offers up a future base for additional `tsc` plugins (e.g. file watching plugins, module resolution plugins).
  - The choice in discriminator (`"type": "transformer"`, for now) is intended to avoid conflicting with existing plugins defined by `ttypescript` or `ts-patch`.
- Plugins would take a form similar to language service plugin, receiving the `ts` object, returning a factory that can be used to create a `CustomTransformers` object. The factory would receive the `Program`, which is required for any type-using plugins.
  - The `ts` object must be a part of the API for similar reasons to the language service plugins; the APIs available within the core compiler are very, very limited, and the bundle is _not_ the same as `typescript.js`.
  - The plugin object entry is passed to the plugin for extra configuration, like LS plugins.
- A new `tsclibrary.d.ts` file is added, describing the limited API within `tsc`. This is similar to `tsserverlibrary.d.ts`.
- When invoking `tsc`, you must pass `--allowPlugins` to opt into executing plugins. When using the API or tsserver, plugins are enabled by default.
  - No sandbox is present otherwise. If you currently use a patching-based method to use plugins, you're already taking matters into your own hand. If you currently use `webpack` or `rollup`, your config files are executable anyway.
  - `--allowPlugins` is also a part of the watch plugin PR.
- If multiple plugins are used, their returned `CustomTransformers` are merged.

There are almost assuredly other things that people want to be able to do with custom transformers or other plugins, however, I believe that we can serve the vast majority of those who want custom transformers with this API, and I do not want to let perfect be the enemy of the good.

In the future, we can expand `plugins` to do more, allowing the customization of other aspects of TypeScript, or simply add on to what custom transformer plugins can do. [Prior research](https://gist.github.com/rbuckton/090b2b5d27079dec5639e79c072a289e) has shown that there are more interested parties, which I believe that we can eventually support.

# An example

With the above proposal, a `tsconfig.json` would look like this:

```jsonc
{
    "compilerOptions": {
        "plugins": [
            { "type": "transformer", "path": "@jakebailey/cool-transformer" }
        ]
    }
}
```

The transformer would look something like:

```ts
import type * as ts from "typescript/tsclibrary";

const factory: CustomTransformersModuleFactory = ({ typescript }) => {
    return {
        create: ({ program }) => {
            return {
                before: [(context) => (file) => {/* ... */}],
            };
        },
    };
};

export = factory;
```

Of course, the details may change.

# Unsolved problems

- It's very unfortunate to have to define _yet another_ random `d.ts` API used for one thing.
  - Can we do better?
  - What if the executables were ESM?
- Like LS plugins, tsc plugins would need to be CJS.
  - Can we instead make use of `import`?
  - How do we do `async`? (This may already be "solved" in a prior WIP for plugins)
- Should we do any sort of performance metric counting for plugins?
  - Like LS plugins, can we just say "no warranty" when plugins are involved?
- How much of a problem is it going to be to need to use the `typescript` namespace passed into the plugin?
  - What happens to existing transformers which mistakenly import `typescript.js`'s API directly, even though that may not have always worked?
  - Will plugins unknowingly use the `typescript.js` API and rely on it, assuming users won't use `tsc`?
  - Should we try and attempt converting our binaries/libraries to ESM first?
- Is the proposed API too factory-y?
- Should we even do this?

# Please give feedback!

If you are currently using transformers, creating transformers, or otherwise, I would appreciate your feedback.

**Comment by jakebailey** [comment]
Preemptive @nonara ping; I suspect you have feedback! 😄

**Comment by timocov** [comment]
Hey @jakebailey,

Thank you for this proposal! I personally like the approach (it feels similar to `ttypescript`), but I have a question - it seems it is not possible to provide options to a "transformer" from the config, right? Would it be possible to allow to define something like `options`/`params`/`config` next to `type` and `path` fields that the compiler will pass as-is (without any transformation/validation) to the transformer function?

<details>
<summary>Quick intro where I came from</summary>

I'm the author of 2 optimization-related transformers:
- https://github.com/timocov/ts-transformer-minify-privates
- https://github.com/timocov/ts-transformer-properties-rename

They both allow you to minify properties of objects (either private or internal) and it is achieved by adding a prefix to "all properties that should be minified" so you can use these prefixes in tools like `terser` or `uglifyjs` to minify i.e. mangle these properties. While the default prefixes (`_private_` and `_internal_`) are used in majority cases, there are few cases when people prefer to use different prefixes for whatever reason (e.g. to sync values in their config files). Thus it would be quite beneficial to have such option.

</details>

Thanks!

**Comment by jakebailey** [comment]
> Would it be possible to allow to define something like `options`/`params`/`config` next to `type` and `path` fields that the compiler will pass as-is (without any transformation/validation) to the transformer function?

Oops, I should have explicitly defined that; yes, any other options are passed along to `create`, just like language service plugins.

```ts
export type CustomTransformersModuleFactory = (mod: { typescript: typeof ts }) => CustomTransformersModule;

export interface CustomTransformersModuleWithName {
    name: string;
    module: CustomTransformersModule;
}

export interface CustomTransformersModule {
    create(createInfo: CustomTransformersCreateInfo): CustomTransformers;
}

export interface CustomTransformersCreateInfo {
    program: Program;
    config: any;
}
```

**Comment by patdx** [comment]
Why wouldn’t it support esm? Is it because the initializing code of tsc is synchronous? It may be nice if it could support esm from the start.

**Comment by jakebailey** [comment]
> Why wouldn’t it support esm? Is it because the initializing code of tsc is synchronous? It may be nice if it could support esm from the start.

Yes, correct, in its current form. The same currently applies to tsserver plugins.

We'd need dynamic import and that doesn't fit nicely into our current architecture. But, I need to go check what was done in the previous attempt as I believe it may have been solved there.

**Comment by samchon** [comment]
It seems great. Hope to see it in near future

**Comment by samchon** [comment]
`TransformerFactory` has been changed to `CustomTransformersModuleFactory`. It seems cool.

**Comment by samchon** [comment]
This is just my opinion - @jakebailey 

## Support both `TransformerFactory` and `CustomTransformerFactory`

Looking at changed codes, I think it may possible to support both `TransformerFactory` and `CustomTransformerFactory`. There're some useful libraries using `ttypescript`, but some of them had stopped maintaining. I think supporting both interfaces maybe helpful for them.



## `--allowPlugins` to be default `true` or just remove it

To use transformation based libraries like my `typia` (or `nestia`), users must configure `tsconfig.json` file with `plugin` property. Configuring the `plugin` property by themselves mean that they have already agreed to using plugin transformation. In my opinion, typing `--allowPlugins` CLI is just annoying thing for users.

I hope it to be default true, or just remove it.

**Comment by jakebailey** [comment]
> Looking at changed codes, I think it may possible to support both `TransformerFactory` and `CustomTransformerFactory`. There're some useful libraries using `ttypescript`, but some of them had stopped maintaining. I think supporting both interfaces maybe helpful for them.

Those transforms are unlikely to work because they import `typescript` directly. The APIs available within `tsc` are very limited, no methods on `Node`, `Type`, etc, and it's very possible for a user to accidentally get into a setup where a plugin may import a different version of typescript than is loaded. These plugins need to be loaded in the language service as well, which for most users is _guaranteed_ to not be TypeScript within the workspace, but instead be the one included in VS Code and therefore cause two different versions of TypeScript to be active in memory and potentially have mismatching behavior.

> To use transformation based libraries like my `typia` (or `nestia`), users must configure `tsconfig.json` file with `plugin` property. Configuring the `plugin` property by themselves mean that they have already agreed to using plugin transformation. In my opinion, typing `--allowPlugins` CLI is just annoying thing for users.

This comes down to the security tradeoffs. In the API, you're already running code, so the flag isn't needed. In the language service, we already load plugins automatically and in VS Code require that you "trust" the workspace before starting `tsserver`. There is no such equivalent for `tsc` and without `--allowPlugins`, this would be the first instance of `tsc` arbitrarily executing code.

---

I should also note that it's possible that this proposal is _not_ implemented at all. There are still strong arguments against doing this sort of thing whatsoever. The use cases may not outweigh the downsides; it may also be the case that this feature would be avoided by most because this means they cannot use any transpilers _except_ TypeScript itself.

**Comment by canonic-epicure** [comment]
> ## `--allowPlugins` to be default `true` or just remove it
> To use transformation based libraries like my `typia` (or `nestia`), users must configure `tsconfig.json` file with `plugin` property. Configuring the `plugin` property by themselves mean that they have already agreed to using plugin transformation. In my opinion, typing `--allowPlugins` CLI is just annoying thing for users.
> 
> I hope it to be default true, or just remove it.

+1 for this, `--allowPlugins` is clearly an overkill. Its not a nuclear rocket launch, where you have to use 2 different keys from 2 persons, rotate them simultaneously etc. If user has specified the transformer in the "tsconfig.json" I believe his/her intention is clear enough.

**Comment by amanteaux** [comment]
Another use case for custom transformer is to provide dependency injection at compile time. This is what is partially done in https://github.com/wessberg/DI-compiler

**Comment by 43081j** [comment]
I've actually pushed for this many times over the years as I've written a fair few transforms, most of which are for codegen stuff (e.g. replacing decorators with equivalent code).

However, I think i'd vote against it these days.

If we open the flood gates here, we're opening ourselves up to inconsistent behaviour and syntax across any two typescript projects - which seems high risk to me.

The usual culprits that already implement a lot of "magic" (frameworks, codegen projects, gql, etc) i'm sure would make use of it and build their own "flavour" of typescript soon enough. Which would lead to fragmentation of an otherwise pretty concrete ecosystem/community.

On top of that, typescript was always meant to be "javascript with types" (i.e. just some annotations we can strip away). That kind of goes away if we open up to custom syntax.

I'd love to be able to simplify some of my builds by this being possible, but i would rather put up with needing my own build script to avoid the above.

**Comment by samchon** [comment]
> ## Summary
> TS 5.2 transform API removed type info.

I'd tried to adjust this TS 5.2 transform API in my llibraries [typia](https://github.com/samchon/typia) and [nestia](https://github.com/samchon/nestia).

However, it was not possible because most methods and properties are erased from `ts.Node` and `ts.Type` in TS 5.2 transform API. By such removes, even checking `number` type became impossible, due to `ts.Type.getFlags()` method has been erased. Another words, I can't support runtime validation feature through TS 5.2 transform API.

I asked the reason why in the related PR, but got answer that it's for performance enhancement (https://github.com/microsoft/TypeScript/pull/54278#issuecomment-1550830073). Unfortunately, it just sounds me like "I removed the seatbelt and fuel efficiency went up." Considering the purpose of the TypeScript project, I think this is nonsense.

@jakebailey I'm sorry for criticizing even from the 1st PR. However, type info is very important feature for me.

**Comment by jakebailey** [comment]
I don't understand the above; clearly all type information is available within tsc, it's just not as convenient as how it is within the wider services API. If we were to do this, we would of course want to ensure that things are accessible; that is the point after all. I mentioned some of this preciously.

I'll also say that I don't understand how what you're saying is accurate when the patching libs still operate within the same bounds as tsc already has. How are they getting the methods, when such methods don't even exist within tsc?

**Comment by samchon** [comment]
@jakebailey As you've said `typescript` is just convenient and everything can be done `typescript/lib/tsclibrary`, I've studied TypeScript codes for a day. During that times, I could find alternative way like below. By the way, the alternative way means to define duplicated code with TypeScript core codes that are not exported.

1. `ts.Node.getSourceFile()` amd `ts.Node.getFullText()` can be manually implemented, but it would better to TypeScript exports `src/compiler/utilities.ts` without `@internal` tag like `getSourceFileOfNode()` function.
2. `ts.Type.isTypeParameter()` can be replaced by `&` operation (`ts.TypeFlags.TypeParameter`)
3. `ts.Type.isLiteral()` also can be replaced by `&` operation
4. `ts.Type.isIndexType()`, `ts.Type.isIntersection()` types are also possible
5. `ts.Type.isArrayType()` and `ts.Type.isTupleType()` are possible by `ts.TypeChecker`

Besides, there are some features that could not implement easily, or impossible. Traveling TypeScript source codes, I can sure that there're much more features that could not be implemented by 3rd party library. I hope you to consider such aspects, and want at least one of them: just allow `typescript` definitions, or export internal compiler and service functions.

- `ts.SourceFile.getLineAndCharacterOfPosition()` has deep dependency on internal functions. Especially about `lineMap` property, it does not exist on the `ts.SourceFile` type due to `@internal` tag. Also, it may possible to copying and pasting same code on my transformer library, but cannot sure it is right behavior or not.
  1. [src/compiler/scanner.ts#L520-L522](https://github.com/microsoft/TypeScript/blob/dddd0667f012c51582c2ac92c08b8e57f2456587/src/compiler/scanner.ts#L520-L522)
  2. [src/compiler/scanner.ts#L480-L486](https://github.com/microsoft/TypeScript/blob/dddd0667f012c51582c2ac92c08b8e57f2456587/src/compiler/scanner.ts#L480-L486)
  3. [src/compiler/scanner.ts#L408-L435](https://github.com/microsoft/TypeScript/blob/main/src/compiler/scanner.ts#L492-L506)
  4. [src/compiler/core.ts#L1276-L1315](https://github.com/microsoft/TypeScript/blob/dddd0667f012c51582c2ac92c08b8e57f2456587/src/compiler/core.ts#L1276-L1315)
  5. [src/compiler/scanner.ts#L475-L477](https://github.com/microsoft/TypeScript/blob/dddd0667f012c51582c2ac92c08b8e57f2456587/src/compiler/scanner.ts#L475-L477)
  6. [src/compiler/scanner.ts#L408-L435](https://github.com/microsoft/TypeScript/blob/dddd0667f012c51582c2ac92c08b8e57f2456587/src/compiler/scanner.ts#L408-L435)
- `ts.Symbol.getDocumentationComment()` and `ts.Symbol.getJsDocTags()` also related with deep internal functions. 
- `ts.Node.getChildren()` has the most complicate relations with internal functions
- About `ts.transpile()` function, I just gave up

**Comment by jakebailey** [comment]
re: https://github.com/microsoft/TypeScript/issues/54276#issuecomment-1551366861

1. Use `ts.getSourceFileOfNode` and `ts.getLineAndCharacterOfPosition`.
2. Use `ts.getSourceFileOfNode`.
3. Use `ts.getSourceFileOfNode`.
4. Check `type.flags & TypeFlags.TypeParameter`.
5. Check `type.flags & (TypeFlags.StringLiteral | TypeFlags.NumberLiteral | TypeFlags.BigIntLiteral)`. `LiteralValue`'s value property is public.
6. This can be checked via `objectFlags`. We could export a helper for this.
7. I'm not convinced `getChildren` should be used in a transform at all. Our transforms do not use this helper. You probably want to be using explicit visit functions. For the properties, call the checker, which has `getPropertiesOfType` and `getAugmentedPropertiesOfType`.
8. I don't think a transform should need `getDocumentationComment`. That's for the language service for displaying in tooltips and such. There are other ways to obtain JSDoc info, too.
9. Check `type.flags & TypeFlags.Union` or `type.flags & TypeFlags.Intersection`.
10. Just access `type.flags`.
11. Just access `type.symbol`.
12. Call `typeChecker.isTupleType`.
13. Call `typeChecker.isArrayType`, `typeChecker.isArrayLikeType`, or `typeChecker.getIndexTypeOfType`.
14. Call `ts.getSourceFileOfNode` and slice the text by `node.pos` and `node.end`.

When you operate within the core compiler, you have to play by the core compiler's rules. These are all things that we do on a regular basis. If there is an API missing, then we can add it.

But pushing FUD on multiple issue trackers is really not helping, in fact, it's mostly convincing me that I shouldn't have made this proposal at all.

**Comment by samchon** [comment]
Okay, I will argue only here issue

**Comment by samchon** [comment]
@jakebailey How to replace `ts.Symbol.getJsDocTags()` method?

Also, In my case, `ts.symbol.getDocumentationComment()` is required for JSON schema definition.

**Comment by jakebailey** [comment]
I suspect via #53627 or the `.jsdoc` property on the declarations.

**Comment by nonara** [comment]
Looks interesting, @jakebailey! Thanks for looping me in. I'll share a few thoughts below.

> wholly replace TypeScript in a user's workflow ... eek!

I understand where you're coming from. I think in a broad sense, I probably wouldn't classify what's being done as solely this.

What we're essentially doing is injecting some additional code into the underlying file to extend its capability. In the case of tsc it's been a little more extreme, in that we have to splice together logic from two modules.

This can also happen in memory, where we're not actually replacing anything, which has been the practice for ttsc and is now integrated into tsp as well.

Moving forward, our methodology is further simplified as simply running transformers on the compiler to add functionality. 

With that in mind, I don't think I'd necessarily agree that this is "eek" (😂), as I think that has some very real value in specific cases. If memory serves, IDEs like JetBrains (at least at one point) were doing similar.

That said, when confronting the question of whether it _should_ be necessary to actually patch ts for something that has been so widely requested and is now widely used, I definitely agree that it makes sense to take a closer look. On the surface, it seems a bit extreme to need to patch ts to do it.

**Should We?**
> "Your scientists were so preoccupied with whether or not they could, they didn’t stop to think if they should" — from the great philosophical opus, "Jurassic Park"

Certainly, if it was simple to fully support, it would be an easy answer, but in this case, you face the trade-off of either sacrificing the reduced size of tsc (presumably also sacrificing some performance), or in offering something which has limitations that prevent many use-cases.

One possible option I might consider would be creating a separate, full binary (eg. tscp.js), which has full plugin support. Obviously, though, this all depends on what your priorities are, and I'm not advocating for that. On a simply surface level, this would ameliorate issues and provide a full solution.

A secondary option is to simply do it how you've proposed, and tsp can extend capability for those who need it.

**Issues with New TS 5**

One quick side tangent, as I think it has been widely misunderstood (understandably) across the community.

The changes made to TS 5.0 were actually not that much of a challenge to accommodate. In fact, esbuild made it a bit easier to identify the surgical points.

Ultimately, there were a couple of main reasons it took me a a while to do. Mainly, I took it as an opportunity to do some major rewrites I had planned for probably two years. This was in part because it was required.

That aside, I've also just been slammed. I did, however, get a working beta out ASAP.

I can understand why people got nervous, but putting it into perspective, the worst that happened was people had to wait a little bit to try the latest and greatest. This is obviously not ideal, and part of the reason for the rewrite is to make it so that doesn't happen moving forward. 

Anyway, I don't want to hijack the thread on that.

**Further Responses**

> A future API that actually allows a consumer to tree shake out just a parser (or similar).

Sounds very interesting! Looking forward to seeing what you come up with.

> with a discriminator like "type": "transformer"

This sounds great. I didn't choose the current config format, and I don't like how it's laid out. 

I was planning on moving plugin config for future tsp outside of tsconfig entirely, but I'd be glad to conform to whatever is officially released.

> A new tsclibrary.d.ts file is added, describing the limited API within tsc. This is similar to tsserverlibrary.d.ts.

I like this approach.

> How much of a problem is it going to be to need to use the typescript namespace passed into the plugin?

This one is interesting. One thing that's very important with writing proper transformers is to use the passed ts instance. Very few transformers actually do this, and it can cause issues in numerous cases.

I think the most important part would be good documentation. 

One possibility (though I don't know if it's worth it or a good idea) would be to "sign" root elements, like a SourceFile and possibly Program, so you could throw if they tried to pass to an incompatible instance.

**ESM**

This part is certainly tricky. I ended up getting esm transformer support to work in tsp (still have to address a memory leak issue) by transpiling to CJS on-the-fly.

**Summing it up**

It's an interesting proposal, and I love the idea of a better config format. Your research and thought process so far is definitely on track!

Looking forward to keeping up with it. 

As for the noise, just a word of encouragement — keep your head up! I know I don't need to tell you, but it's a complex subject with a lot of specialized knowledge, which broadens the field of those who'll misunderstand. I certainly know how frustrating the feedback can be when people don't fully grasp the issues, especially when that's matched by boldness and curt expression.

That said, I can't begin to imagine what you all have to deal with on a daily basis. I've seen some threads!

I think I can speak for a good portion of the silent majority who appreciate the efforts to investigate this further and to come up with a fully well-thought out determination on approach, and I appreciate the invitation to weigh in.

**Comment by jakebailey** [comment]
> With that in mind, I don't think I'd necessarily agree that this is "eek" (😂), as I think that has some very real value in specific cases. If memory serves, IDEs like JetBrains (at least at one point) were doing similar.

Truthfully, my main motivation here is that I would really like to eliminate all of the external need to patch the TypeScript package; the fact that people do this limits what we can do in our package, e.g. we can't think about doing any sort of minification because doing so means that either it'll break `.patch` files (which work on lines; e.g. yarn's patching), potentially mangle variable names (which a syntactic transform may break with), and remove comments (e.g. the hack that prettier uses to extract just the parser).



> Certainly, if it was simple to fully support, it would be an easy answer, but in this case, you face the trade-off of either sacrificing the reduced size of tsc (presumably also sacrificing some performance), or in offering something which has limitations that prevent many use-cases.

Size wise, I'm not worried. It's only 200KB or so in `tsc`, just the effect of not tree shaking due to any API potentially being used by a plugin. See the size report on my WIP PR: https://github.com/microsoft/TypeScript/actions/runs/4997656840#summary-13534656632



> One possible option I might consider would be creating a separate, full binary (eg. tscp.js), which has full API support for plugins. Obviously, though, this all depends on what your priorities are, and I'm not advocating for that. On a simply surface level, this would ameliorate issues and provide a full solution.

Another option to "solve" this may be to try and restructure our API to make it easier to build custom CLIs that enable more features like plugins. Right now, I think that is challenging, but I think it could be very possible. I already want to try and ESM-ify the package.



> One quick side tangent, as I think it has been widely misunderstood (understandably) across the community.
> 
> The changes made to TS 5.0 were actually not that much of a challenge to accommodate. In fact, esbuild made it a bit easier to identify the surgical points.

Yes, totally. Though, I can't say I'm super happy overall that people are using the `// src/compiler/...` comments for patching...

**Comment by jakebailey** [comment]
@samchon In order to correctly support the plugin model in this proposal, you need to consistently use the correct API across your project, passing the `ts` module provided by the `create` call around or just setting it to some global. This is a restriction already imposed on language service plugins; you just cannot mix and match the various bundles as they are all different, contain different things, have different classes which will fail `instanceof`, etc. You always need to be using the API object passed in by the plugin system, not any import.

**Comment by samchon** [comment]
Ah, understood. Have to use `typescript` instance gotten from `CustomTransformersModuleFactory`.

**Comment by samchon** [comment]
> @samchon In order to correctly support the plugin model in this proposal, you need to consistently use the correct API across your project, passing the `ts` module provided by the `create` call around or just setting it to some global. This is a restriction already imposed on language service plugins; you just cannot mix and match the various bundles as they are all different, contain different things, have different classes which will fail `instanceof`, etc. You always need to be using the API object passed in by the plugin system, not any import.

In my case, `typia` is used by both transformation, generation and library mode. As you know, transformation mode means using transform API. Generation mode means generating duplicated TypeScript files with transformation. Library mode means used by other library by importing.

By the way, types of `typescript` can't be assigned to `typescript/lib/tsclibrary`. In that case, must I make duplicated codes to support TS 5.2 transform API? Or is there any other way?

**Comment by jakebailey** [comment]
That I'd need to think about; that's part of my "it's unfortunate to have yet another declaration" problem. But theoretically, you could scope your project down to only use what's available in `tsclibrary` and `typescript` should be a superset that is assignable down to `tsclibrary`.

I do want to get feedback on this proposal from others, so I don't want to clog this thread up with specifics of your project, however.

**Comment by nonara** [comment]
@jakebailey 

All great points. I take a pretty hard stance on this, which is that regarding anything not explicitly deemed public API, I have no reasonable expectation of this not changing at any point.

It's great that you guys consider broader tooling, and I think that's important, especially in situations like yarn or IDEs.

What we're doing, in my mind, has always been a somewhat experimental status way to do things you couldn't otherwise do, which is really what my vision for the future is. For example, it would provide an excellent way to make portable patches to test new features to the language in a proposal. 

This does, of course, carry with it the expectation that it's up to the patch authors to keep up with new versions. We all have to do this with TS changes anyway, but there will definitely be more risk on this side of things. That should be understood, and I think it makes sense for me to add a better disclaimer to that effect in our documentation.

With that said, I do get that this can add some additional stress or perception of constraint on your side, which I don't want to be the case.

> we can't think about doing any sort of minification because doing so means that either it'll break .patch files

This does underscore my thinking on it making sense to have a better approach to patching TS for those of us who need to do it. Ideally, it would be with a single tool, which is the hope.

The method I have in place now slices the file into its original pieces, and it essentially has them as memoized SourceFiles, which you can transform. This performs very well, and it makes it easy to modify things using transformers. This approach is also much more sturdy and requires less maintenance than `.patch ` files.

Ultimately, the rewritten tsp will allow modules to easily take this info and transform exactly what you need. It also leverages caching well and allows in-memory changes, which should work well for IDEs, yarn, etc. I expect this should not only require less maintenance, but it should also be faster for them (after the first run). The trick is just ensuring it's planned out very well.

> I can't say I'm super happy overall that people are using the // src/compiler/... comments for patching...

To clarify, I didn't expect that these would remain present. Prior to this, our boundary detection was more flimsy, using a regexp pattern, so it made sense to use it, but I had the expectation of possibly changing it later.

But your point on not being able to do things like minification is a very good one.

Two immediate options I can see would be to either publish a second package with the uncompressed versions or to include the uncompressed binaries in a subdir (if the added filesize was acceptable)

**Potential Proposal**

I think ultimately, the hope of entirely eliminating the need to modify typescript is not going to be realizable. 

The cases in which it needs to be done are going to be rare and often crazy use-cases, but those often happen to pop up in places like yarn or major IDEs, which unfortunately mean that rarity doesn't make them rare in the wild. No doubt, more will emerge in time, which would likely push beyond the bounds of whatever generator may be built.

However, I also don't like the fact that this is a pain-point which hinders further progress on the development of the compiler.

I wonder if instead it makes sense to work out a standard, so we can have a more formalized pathway, that gives us a few guarantees we can hold to and alleviates the issues for your side.

I do think having a tool which allows augmenting TS via transformation-based patches is a great approach, with advantages over binary patches, including lower maintenance to keep up. The major hurdle there was performance, but our current approach handles that very nicely.

As far as a standard, the idea would be to do something which provides a few basic guarantees and – more importantly – that takes off the constraints from your side.

Here is a general example:

- Add a second directory called `lib-full`, with just the js library binaries that remained un-minified.
- These contain a comment in a special format which delineates between files (not tied to esbuild)
- Each release could have a short description of _notable_ changes to the internals, like files renamed, functions moved, etc.

I think a few things like this would make it easier for us and patch developers to keep up, but I welcome feedback on that or feasibility, especially of the third.

Again, if we were to consider this route, I'd stress to everyone (to posterity if this becomes a broader discussion) that the goal here is to make things _easier for the TS team_ first and foremost, removing constraints on them from improving the compiler.

Second, standardizing a format for a tool that can be used across the board for augmenting the compiler will be beneficial to the user community.

**Comment by treybrisbane** [comment]
@marcj heads up - this may have implications for Deepkit. 🙂

**Comment by jakebailey** [comment]
@uasan No, I don't think so. That sounds very unrelated to me.

**Comment by canonic-epicure** [comment]
I'd like to address the concern about fragmentation of TS ecosystem after implementing this proposal. 

Lets look at Haskell. They have ["language extensions" ](https://wiki.haskell.org/Language_extensions), which can add new features to the language, up to modifying the grammar and adding new syntax. Historically, there is a Haskell 2010 standard and now they are discussing the Haskell 2020 standard which is basically a Haskell 2010 + a set of commonly used language extensions, which have proven to be useful.

All those experiments with the language were happening in the user-space, the core team didn't have to spent time on those. And then useful ones are becoming a new standard. Users have to opt-in for every extension, but all extensions 
 are interoperable, because they all compiles down to the intermediate language.

This is an example of how a language ecosystem can evolve, requiring minimal effort from the core team and utilizing the full creativity power of the community. You can't move forward by standing on the same place.

TS transformers are much simpler than that, they are supposed to only modify the JS output. As in the Haskell case, there is an intermediate representation which is shared by all code (JavaScript). Users have to opt-in for every transformer, there's no risk of breaking any existing code.

I believe restricting the language diversity is something that limits the TypeScript usage in the long run.

**Comment by dvins** [comment]
> All those experiments with the language were happening in the user-space, the core team didn't have to spent time on those. And then useful ones are becoming a new standard. Users have to opt-in for every extension, but all extensions are interoperable, because they all compiles down to the intermediate language.
> 
> This is an example of how a language ecosystem can evolve, requiring minimal effort from the core team and utilizing the full creativity power of the community. You can't move forward by standing on the same place.

Well said @canonic-epicure!

**Comment by mindplay-dk** [comment]
@jakebailey please don't be discouraged 🙂

Your efforts to push for something, anything to happen in this area is greatly appreciated! 🙏

I wouldn't personally expect TS to directly adopt any transformer format previously defined by the community - I understand this might create more work for the community already invested in certain transformer formats, and I do think there's some risk of fragmentation if existing transformers continue to use community formats not officially supported by TS. But I think that's the cost of making progress - and as @nonara said, priority is to avoid creating problematic constraints for the compiler team.

Thank you @nonara and @canonic-epicure for your measured comments 👍

**Comment by mrkev** [comment]
Should some solution for versioning be included at the plugin level, or would `peerDependencies` in `package.json` suffice? 

I'm wondering if `tsc` would want to know what version of typescript its plugins expect. If the only utility for this is warning about unsupported versions, perhaps it's redundant though.

**Comment by mindplay-dk** [comment]
@mrkev TS is semantically versioned - so I'd assume any change to a public API is considered a breaking change.

The general problem with plug-ins is [well known](https://nodejs.org/en/blog/npm/peer-dependencies#the-problem-plugins) - and it's not really `tsc`'s job to manage dependencies.

**Comment by jakebailey** [comment]
> TS is semantically versioned - so I'd assume any change to a public API is considered a breaking change.

TypeScript is not semantically versioned; we just bump from 5.1 to 5.2, and when we hit 5.9 the next version will be 6.0. For the public API, we try and not break things, including using a special deprecation system to provide an early warning. But that system is not included in tsc (historically).

**Comment by sosoba** [comment]
IMHO plugin should not have a dependency on `typescript`. The actual reference is passed in the arguments:
https://github.com/cevek/ttypescript/blob/835c77bfe1a84bdafc7e223ece93f8a5a949243c/packages/ttypescript/src/PluginCreator.ts#L47-L51

**Comment by mrmckeb** [comment]
I haven't left feedback yet, but as far as I understand, this would allow plugins like ([`typescript-plugin-css-modules`](https://github.com/mrmckeb/typescript-plugin-css-modules) to now emit that type information outside of the API (and IDE).

If my understanding is correct, I think this will lead to a landslide of plugins.

I'm more than happy to put some time into prototyping around this too. I'm in Australia, but can always make time for US-friendly chats too.

My thoughts on the proposal:
- I'd like to see `--allowPlugins` available as a `tsconfig.json` option.
- Async support is more important than ESM from my experiences, but ESM support would be great, especially because a lot of projects are moving to ESM-only.
- Native TypeScript support would be cool, as everyone likes no-build projects.
- I think it would be great to capture plugin performance metrics, but that could be something that shipped later.

> Should we even do this?
Definitely ;)

**Comment by jakebailey** [comment]
> * Async support is more important than ESM from my experiences, but ESM support would be great, especially because a lot of projects are moving to ESM-only.

Can you clarify what you mean by "async"? The only place where async and ESM apply is in async loading of plugins (which is feasible though challenging, compared to making anything else async which is extremely far off).

**Comment by mrmckeb** [comment]
> Can you clarify what you mean by "async"? The only place where async and ESM apply is in async loading of plugins (which is feasible though challenging, compared to making anything else async which is extremely far off).

Yes, sorry. The plugin I maintain (and you've helpfully contributed to) makes use of a number of dependencies that have been moving to async-only APIs. As the methods we're proxying aren't async, we can't make use of those APIs. Mainly `getScriptSnapshot` I think.

**Comment by jakebailey** [comment]
TypeScript internally is hugely massively far off from incorporating anything but the most basic async stuff, e.g. at most, I _think_ we could do enough async to do a dynamic import of ESM (which we _kinda_ do for language service plugins). The rest... I'm pretty sure is going to have to be synchronous. But, I haven't had time to go reread previous attempts at plugins to know the extent of async there.

**Comment by samchon** [comment]
@jakebailey 

In NestJS case, it guides user to compile TypeScript source codes through `nest build` command instead of `tsc`. I'm testing in NestJS, but it is hard to advance because of that. Can you give another option instead of `--allowPlugins` argument? At least `allowPlugins` configuration can be done in `tsconfig.json` file, it would much better I think.

**Comment by jakebailey** [comment]
I'm not sure what you mean; if someone is wrapping `tsc`, then they can pass the flag or use the API which doesn't have the flag.

The flag is there so that `tsc` is not arbitrarily executing code without permission. I personally would not be thrilled to check out an untrusted TS issue repro and be unable to compile it without potentially harming my system due to a malicious plugin.

**Comment by samchon** [comment]
Well, unfortunately `nest build` is not delivering arguments to `tsc`. 

If no plan to support alternative option, I'll just send a PR to NestJS with explanation. 

Thanks for replying.

**Comment by jakebailey** [comment]
I'll reiterate that this issue is only a proposal. We have not committed to this API, or even the feature at all.

I would not recommend working on making your projects compatible with this proposal or to send PRs to other projects until we actually know it's going to be in a release.

**Comment by lorenzodallavecchia** [comment]
> The flag is there so that `tsc` is not arbitrarily executing code without permission.

I'm not convinced about the "untrusted code" argument. 
Yes, the concern is real, but many build tools support plugins, like webpack, eslint, karma... even npm itself may install any declared package and use it in the `scripts` field of `package.json`, even automatically at install time.
How is `tsc` different?

**Comment by jakebailey** [comment]
Those to me are all fine reasons to not be against plugins on the whole, and is in fact why I'm not worried about sandboxing. That being said, here are two such problems with letting plugins be enabled without a flag:

- From the TS bug reproduction and breaking change detection point of view, we already rely on the ability to clone repos and `npm install --ignore-scripts` to run `tsc` on them. Either by hand or using automation (those top100/user tests/etc which you see us running on PRs). As I said before, the invariant of "running tsc is a safe operation" would be pretty unfortunate to give up.
- If you're using VS Code, it already prompts you to to say that you trust your workspace before code is executed. There's no equivalent for tsc at the command line, because it doesn't do CLI plugins... until this proposal.

**Comment by canonic-epicure** [comment]
> Yes, the concern is real, but many build tools support plugins, like webpack, eslint, karma... even npm itself may install any declared package and use it in the scripts field of package.json, even automatically at install time.
> How is tsc different?
Exactly, the whole landscape of web-dev tools uses plugins w/o this sandboxing concern.

**Comment by jakebailey** [comment]
> There's a bigger interest there in letting app tooling understand the trust level of a workspace: could VS Code or Webstorm tell tsc that the workspace is trusted?

VS Code has a trust model in place (unless you have disabled it). When a workspace is untrusted, tsserver is not loaded at all thanks to LS plugins (and possibly the tsdk setting). Theoretically we could improve that. But that's a different issue that seemingly nobody has complained about as of yet.

**Comment by zaguiini** [comment]
Is there a way to recreate the program through a custom transformer plugin?

Use-case: I'd like to add new files to the program and these files need to be included in the type-checking phase.

It's not only about the emit phase but about changing the structure of the program. Right now I'm using @nonara's `ts-patch` with a program transformer. [Here's what the transformer looks like](https://github.com/zaguiini/typescript-virtual-barrel/blob/master/packages/compiler-plugin/src/index.ts), heavily inspired by [Ron's reply to this discussion](https://github.com/nonara/ts-patch/discussions/29).

**Comment by canonic-epicure** [comment]
Hm.. This makes this proposal less useful and definitely not a replacement for `ts-patch`.

Why is the idea of [macroses](https://stackoverflow.com/questions/62501505/implementing-macros-typescript-transformer) (AST => AST transformer, before the type check) perceived so negatively by the TS team? Afaik, Rust and Go use such macroses widely and nothing bad happens in their world.

In fact, in Rust, macroses are `TokenStream -> TokenStream` so, for example, one can [safely embed sql right into the Rust code](https://doc.rust-lang.org/book/ch19-06-macros.html#function-like-macros). And nothing bad happens, Rust community is not fragmented because of that. 

So how TypeScript is different from Rust, that the idea of macroses is unacceptable in it?

**Comment by zaguiini** [comment]
>This proposal is strictly emit only (transformers)

How is it different from the current transformers, then? The only new perk is access to the program, but we're essentially limited to the same capabilities.

>code generated by external tools

I see it as a bad thing that you need to run external tools, before or after the compiler runs, to get your source/output transformed. It breaks bundlers, creates awkward ergonomics through `pre` and `post` NPM scripts, and is often error-prone because the user more often than not might simply forget to run something. Even if they don't forget, it's more mental workload/a new mental model to remember to run a specific step in order to successfully compile a project.

Pre-emit transformers could easily be a win. Take [typescript-virtual-barrel](https://github.com/zaguiini/typescript-virtual-barrel) as an example: previous solutions needed an executable running side-by-side with the compiler every single time you're working on your project. This tool completely removes the need for that: you just clone the project, install your deps and you're good to go. Any new changes to the files will automatically regenerate the barrel file because the program is being modified directly.

**Comment by michaelboyles** [comment]
I've written a couple of typescript transformers:
- https://github.com/michaelboyles/jsx-conditionals
- https://github.com/michaelboyles/redcr

The minimal proposal looks mostly good to me. I'd echo thoughts that `--allowPlugins` seems like overkill. I'd either have `--disallowPlugins` or nothing.

One of the things people I've shared these with have said to me is basically "seems cool but setting up the transformer sounds like a faff", and this would solve that.

I'm looking forward to seeing this make it into a release. Thanks for the proposal and the work on the implementation so far.

**Comment by tobias-tengler** [comment]
I'm wondering whether this proposal would allow to clean up code using [Relay.js](https://relay.dev/). Currently product code would look like this:

```ts
import { NameOfQuery } from "./__generated__/NameOfQuery.graphql";

const query = graphql`
  # NameOfQuery will be the name of the generated file
  query NameOfQuery {
    # ...
 }
`

const Component = () => {
  const data = useLazyLoadQuery<NameOfQuery>(query, {});
  // ...
}
```

Relay.js has a compiler step that transforms the `graphql` annotated string literals into type definitions. Using Babel.js the `graphql` definition is replaced with an import of a JS object that is generated alongside the type definitions. This JS object is then passed to Relay.js functions.

Would this proposal allow to emit code that types the return of `graphql` as the type imported from `./__generated__/<NameInGraphQLStringLiteral>.graphql`?

```ts
// query would now be typed as NameOfQuery
const query = graphql`
  query NameOfQuery {
    # ...
 }
`

const Component = () => {
  // This function can now infer the generic argument from the query variable
  const data = useLazyLoadQuery(query, {});

  // ...
}
```

This would allow to get rid of sooo much boilerplate and the need to keep imports in sync with the name inside of the `graphql` string literal.
Doing this in the emit step would help with not producing any errors when running `tsc`, but you would still need a TypeScript LSP plugin alongside this to not make an IDE/Editor complain, right?

**Comment by michaelboyles** [comment]
@tobias-tengler All this proposal is for is to directly expose a feature that already exists. You can write a TS transformer right now to test whether it can do what you want. See the section "How do users currently use custom transformers?" in the issue. There's at least 5 ways to do it! See [this handbook as well](https://github.com/itsdouges/typescript-transformer-handbook)

I don't think I fully understand everything in your example, but I don't think it can achieve what you want. The TS compiler phases are, in order: parse, bind, check, **transform**, emit. Type checking has already finished before the code is transformed, and it seems like you'd want that to be the other way around. Also yes, it looks like you'd need an IDE plugin.

**Comment by samchon** [comment]
> 

What a nice docs repo, thanks for link

**Comment by samchon** [comment]
> @tobias-tengler All this proposal is for is to directly expose a feature that already exists. You can write a TS transformer right now to test whether it can do what you want. See the section "How do users currently use custom transformers?" in the issue. There's at least 5 ways to do it! See [this handbook as well](https://github.com/itsdouges/typescript-transformer-handbook)
> 
> I don't think I fully understand everything in your example, but I don't think it can achieve what you want. The TS compiler phases are, in order: parse, bind, check, **transform**, emit. Type checking has already finished before the code is transformed, and it seems like you'd want that to be the other way around. Also yes, it looks like you'd need an IDE plugin.

What a nice docs repo, thanks for link

**Comment by tobias-tengler** [comment]
> @tobias-tengler All this proposal is for is to directly expose a feature that already exists. You can write a TS transformer right now to test whether it can do what you want. See the section "How do users currently use custom transformers?" in the issue. There's at least 5 ways to do it! See [this handbook as well](https://github.com/itsdouges/typescript-transformer-handbook)
> 
> 
> 
> I don't think I fully understand everything in your example, but I don't think it can achieve what you want. The TS compiler phases are, in order: parse, bind, check, **transform**, emit. Type checking has already finished before the code is transformed, and it seems like you'd want that to be the other way around. Also yes, it looks like you'd need an IDE plugin.

Thank your for your answer and input :) 

Very contrived I basically want to be able to extend TypeScript in a way that would allow to dynamically look up a type in another file to use as return type of a function, based on the inputs of the function call. I think there are many other cases where something like this would be useful.
Do you know of a proposal for a feature or even an existing feature like this?

**Comment by lorenzodallavecchia** [comment]
> Very contrived I basically want to be able to extend TypeScript in a way that would allow to dynamically look up a type in another file to use as return type of a function, [...]

A possible problem here is that TypeScript would not know about that additional file and, under incremental compilation, may not recompile the main file if you change the imported one. Am I correct?

**Comment by rbalicki2** [comment]
I would like to add an additional use case here. I'm the author of Isograph (https://github.com/isographlabs/isograph). One workflow is:
* write `iso` literal (thing `graphql` literal)
* compiler generates files
* babel transforms that `iso` literal into a `require(path_to_generated_file)` statement

This is great, and it works, but the types are lost, and users are required to import the generated file for types to work:

```
// No types inferred
const {data} = useLazyReference(iso`Query.home_page`, {})

const HomePageEntrypoint = import '../__isograph/Query/home_page/entrypoint'
const {data} = useLazyReference<typeof HomePageEntrypoint>(iso`...`, {})
```

In this workflows (and others in Isograph), the ability to have typescript understand the results of the babel transform would substantially improve the DevEx!

**Comment by TylerOliverMCG** [comment]
Hi all! I was looking into this issue because it's common practice with vite to use @/ as an alias to the src folder. However, as far as I can tell, doing this in typescript when building for a library leads to broken paths in the emitted files. It looks like this issue would result in the ability to include a transformer to handle the above file path resolution.

If this issue is still active, what can I do to help it continue forward? The thread states the goal was 5.2.0, but we're well past that release and [this pr](https://github.com/microsoft/TypeScript/pull/54278) is still WIP. Very interested in resolving this issue, or any solutions that don't involve wrapping Typescript to resolve a path alias :)

**Comment by jakebailey** [comment]
Path aliases are not recommended to be used in that way (in fact, we now explicitly say not to do so in the docs). `package.json` import aliases fit that need but do work at runtime, though I'm not quite sure why you need the code to work at runtime if you're using vite (a bundler).

I would not pursue this issue as a solution to path aliases being emitted as-is.

**Comment by TylerOliverMCG** [comment]
Don't want to derail this thread, but I'm building a Vue Component library and then looking to emit typing for non-component entities (composables, utilities, etc). Those utilities reference shared type information from our types folder, and we're using a path alias to reduce the overhead. Runtime works great (Storybook resolves all the aliasing), and the local builds look good. When we build the library, the emitted type files still contain the unresolved paths (@/types, etc).

Is the intended workflow then that we just never use a path alias in Typescript for libraries? I'm re-reading [this documentation](https://www.typescriptlang.org/docs/handbook/modules/guides/choosing-compiler-options.html#im-writing-a-library) but struggling to understand what's incorrect about what I'm doing in concept. Or is there some documentation I'm missing about how to sync up my tsconfig with the vite bundler so that vite resolves the aliasing before type information is built. 

Thanks so much for responding, even knowing that I'm on the wrong path is super helpful :)

**Comment by jakebailey** [comment]
If you're using a bundler, or those paths actually do exist at runtime, then that can be fine; see https://www.typescriptlang.org/tsconfig#paths.

But, for shortening import specifiers, import maps do the same thing without needing to "lie" to TypeScript: https://nodejs.org/api/packages.html#subpath-imports

**Comment by M-jerez** [comment]
I would avoid all the pain and not open the floodgates to transformers. 
Instead just give us a way to reflect type information. From there any othr tools can be done. 

E.i:

```ts

interface User { name: string, age: number}

const myUser: User = {name: 'John', ager: 33};

// `reflec` would be a nue function provided by the compiler and the js output would be an object containing type metadata. 
const myUserType = reflect<typeof myUser>();

Console.log(myUserType.properties.name.kind); // string
```

Just sayin 😉🚀

From there people could import code an use reflection for things like generating code, internationalisation, etc, etc.

**Comment by michaelboyles** [comment]
@M-jerez I don't see how that's productive or on-topic for this issue. The TS team have already said they won't do that: https://github.com/microsoft/TypeScript/issues/3628#issuecomment-1634916613

Transformers have already been implemented. The decision is purely whether to expose them, and how. The decision is not to bin them in favour of a feature that's already been declined.

People have already implemented [reflection](https://github.com/Hookyns/tst-reflect), and arguably the most relevant use-case for reflection, [runtime validation](https://github.com/GoogleFeud/ts-runtime-checks)... They're both implemented using custom transformers. Resolving this issue would make it easier to use both of those libs.

**Comment by M-jerez** [comment]
@michaelboyles this is relevant as I kind of agree with this comment https://github.com/microsoft/TypeScript/issues/54276#issuecomment-1551013832

Transformers opens the gates to getting potatoes instead javascript as output. Nowadays they are rarely used but if they become commonly used that would be a problem (in my opinion). 

If you keep the data at runtime you could solve the same problem as transformers using another extra compilation step + runtime data types.
Anyway I think there are  drawback and advantages to each option.

**Comment by marcj** [comment]
> Transformers opens the gates to getting potatoes instead javascript as output. Nowadays they are rarely used

@M-jerez transformers are very commonly used, they are all over the place, even without people knowing: React with JSX, Angular with their decorators that emit runtime data for dependency injection, NestJS with legacy decorators, all these down-level transformers to support new JS features, and many more. Transformers are the backbone of modern web development and probably 99% of all TS rely on them, most unknowingly. Making them available in a more straightforward "official" way would only be a natural next step.

**Comment by sorenbs** [comment]
Prisma is a Typescript ORM with 3M monthly downloads. We have long wanted to implement LINQ-style queries, and this feature would enable us to deliver that feature with excellent DX. We'd love to prototype something on a pre-release version if that is helpful.

**Comment by samchon** [comment]
> Prisma is a Typescript ORM with 3M monthly downloads. We have long wanted to implement LINQ-style queries, and this feature would enable us to deliver that feature with excellent DX. We'd love to prototype something on a pre-release version if that is helpful.

https://github.com/samchon/typia/blob/f1688280397fbc1fdda9a52f7056a519ba537ab2/package.json#L117

**Comment by sorenbs** [comment]
Thanks @samchon! 

> By the way, prisma is planning support safe raw query expression validated by the compiler?

Yes - that's what we would like to do. If we can do it in a way that delivers excellent DX. We do not want to push specific bundlers or something like ts-patch on our users.

**Comment by alshdavid** [comment]
It would be great to also have support for transformers that could run before the core TypeScript ones do - enabling preprocessor steps (like macros).

Would the LSP consider transformer plugins or would there be an alternate pathway for adding support for custom syntax?

**Comment by jakebailey** [comment]
I'm going to retract my proposal; I don't think this is the direction we're going to go, especially in the context of the native port we just announced.

https://devblogs.microsoft.com/typescript/typescript-native-port/

**Comment by samchon** [comment]
The new native compiler, hope it to support the transformer API too.

**Comment by mindplay-dk** [comment]
> The new native compiler, hope it to support the transformer API too.

@jakebailey any idea if it will?

there is a fairly large ecosystem of compiler extensions via `ttsc` etc. - is there any chance of preserving compatibility (perhaps finally committing to making it official) or will all of this need to be rewritten in Go as well?

**Comment by jakebailey** [comment]
You should really comment on one of the threads over in that repo. This PR is not a great place for long-term discussion.

**Comment by nonara** [comment]
@mindplay-dk You can keep an eye on this thread for our side of it: https://github.com/nonara/ts-patch/issues/181

Short version is, as of this point, the plan is to keep up with the changes.

From what I understand, things are still up in the air on what the Compiler API will look like an what parts of the new compiler will be in / support JS. 

The big thing that we'll need to know is whether JavaScript transformers will still be supported by the new Compiler API. I'll be keeping an eye on things, and I know Anders said the TS team is working with popular tooling creators to ensure there's minimal negative impact to existing tools.
